### DISCUSSION TITLE
Complete Inline Javascript Example

### DISCUSSION THREAD

**Discussion by alandefreitas** [livewire/livewire#1898] [2. Show and tell]
Hey guys. Thanks for the library and the community. I'm a newbie here, and it's great to see that there's such a great community around it already.

The library is so useful and the documentation is great. Everything just works. It also saves me a lot of time because I have a website but web programming is not my main thing.

I'm writing this post because I couldn't find a **complete** example of how to make inline javascript work in a stable manner without some other external libraries like AlpineJS. I've read all over this forum, the documentation, the troubleshooting section, and I can't find anything. I found a few snippets here and there that helped but no robust solution.

What I'll try to do here is provide complete example of inline javascript integration similar to the examples the documentation includes for AlpineJS. Kind of a missing tutorial. This is not a perfect tutorial because I'm trying to figure it all out by myself, but I hope it can help other newbies like me so they don't have to go through all that again. My snippets are not going to be very elegant, but I was hoping I could get some feedback on these points so I'll update this post accordingly. 

If this is not the appropriate place for this kind of post, please let me know.

# The component context

There are three situations that don't work the same and it seems hard to write a component that works for all of them:

1) Parent components
2) Nested components
3) Nested components that will be rendered later

The intention is to find a strategy that works in all of these cases. 

This is very important because (it's my perception that)

1) people will rarely write a website that doesn't depend on an external javascript library here and there, 
2) I assume inline javascript in components are compatible with the whole reason to use livewire, which is to avoid dealing with another javascript library like vue.js, and 
3) it seems like the documentation makes it seem like this is much simpler than it really is, and this was very confusing to me. Everything was so smooth up to this point and then I just felt stupid.

As far as I know, I can't seem to find a complete example that works everywhere.

# The complete example

To reproduce the problem, I'll try to create a hello world example and show what happens in each case so that when we can get this to work in a robust manner maybe this can serve as a kind of tutorial to other people.

## Views

A route for our hello world in `web.php`:

```php
// Snippet 1: routes/web.php
Route::get('/helloworld', function () {
    return view('helloworld');
});
```

Our view will include the simplest components possible for the three situations I described above:

```html
{{-- Snippet 2: resources/views/helloworld.blade.php --}}
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Hello Livewire + Inline Javascript</title>
    <link rel="stylesheet" href="https://unpkg.com/easymde/dist/easymde.min.css">
    <livewire:styles/>
</head>
<body>
<h1>Hello Livewire + Inline Javascript</h1>
<div>
    {{-- Parent component --}}
    <livewire:hello-world.markdown-editor key="parent_editor"/>
    {{-- Nested component --}}
    <livewire:hello-world.nested-markdown-editor key="nested_editors"/>
</div>
</body>
<script src="https://unpkg.com/easymde/dist/easymde.min.js"></script>
<livewire:scripts/>
</html>
```

It can't get simpler than this. 

Note that `<script src="https://unpkg.com/easymde/dist/easymde.min.js"></script>` is there because I don't even want to deal with the problems with `@once @push('scripts')` and nested components that will be rendered later.

## Components without inline javascript

Now to the components:

```bash
# Snippet 3: from ./
php artisan make:livewire HelloWorld/markdownEditor
php artisan make:livewire HelloWorld/NestedMarkdownEditor
```

The markdown editor component will be the simplest possible:

```html
{{-- Snippet 4: resources/views/livewire/hello-world/markdown-editor.blade.php --}}
<div>
    <div>
        <textarea wire:model="text"></textarea>
    </div>
    <div>
        {{ $text }}
    </div>
</div>
```

and the component logic:

```php
<?php
// Snippet 5: app/Http/Livewire/HelloWorld/MarkdownEditor.php

namespace App\Http\Livewire\HelloWorld;

use Livewire\Component;

class MarkdownEditor extends Component
{

    public $text = "# Header\nInitial Markdown";

    public function render()
    {
        return view('livewire.hello-world.markdown-editor');
    }
}
```

Also the simplest example possible. It all works perfectly and demonstrates why livewire is such a success.

![68747470733a2f2f666f72756d2e6c61726176656c2d6c697665776972652e636f6d2f75706c6f6164732f64656661756c742f6f726967696e616c2f32582f652f656431663365643138666138393433373366376264336534353931613634663264393563653731372e706e67](https://user-images.githubusercontent.com/5369819/97091360-d5135080-1611-11eb-9b21-562ea6657296.png)

and 

![](https://forum.laravel-livewire.com/uploads/default/original/2X/5/5a378b177210ddc99315df1da064f8e6b5a090bf.png)
![68747470733a2f2f666f72756d2e6c61726176656c2d6c697665776972652e636f6d2f75706c6f6164732f64656661756c742f6f726967696e616c2f32582f352f356133373862313737323130646463393933313564663164613036346638653662356130393062662e706e67](https://user-images.githubusercontent.com/5369819/97091365-da709b00-1611-11eb-9525-bcb82887cab0.png)

Now to the nested components. We have nested components that are rendered right away and nested components that are rendered later, like when we have tabs, etc. The shortest code to emulate that will [defer loading](https://laravel-livewire.com/docs/2.x/defer-loading) the second subcomponent:

```html
{{-- Snippet 6: resources/views/livewire/hello-world/nested-markdown-editor.blade.php --}}
<div>
    <div>
        <livewire:hello-world.markdown-editor key="nested_editor_1"/>
    </div>
    <div wire:init="loadEditor">
        @if ($readyToLoad)
            <livewire:hello-world.markdown-editor key="nested_editor_2"/>
        @endif
    </div>
</div>
```

And then the simplest logic possible on the component:

```php
<?php
// Snippet 7: app/Http/Livewire/HelloWorld/NestedMarkdownEditor.php
namespace App\Http\Livewire\HelloWorld;

use Livewire\Component;

class NestedMarkdownEditor extends Component
{

    public $readyToLoad = false;

    public function loadEditor()
    {
        $this->readyToLoad = true;
    }

    public function render()
    {
        return view('livewire.hello-world.nested-markdown-editor');
    }
}
```

And again, we see why livewire is so great:

![68747470733a2f2f666f72756d2e6c61726176656c2d6c697665776972652e636f6d2f75706c6f6164732f64656661756c742f6f726967696e616c2f32582f302f306433613930306133306533333134633263353632313832633037623164356230383465626138332e706e67](https://user-images.githubusercontent.com/5369819/97091369-e0ff1280-1611-11eb-969d-05971f9085b2.png)

Now we can try a few things to make this textarea work as a real markdown editor.

## Components with Inline Javascript: Literally following the documentation

This is the snippet the documentation provides for inline javascript:

```html
<div>
    <!-- Your components HTML -->

    <script>
        document.addEventListener('livewire:load', function () {
            // Your JS here.
        })
    </script>
</div>
```

If we follow that literally for our markdown editor, we *could* interpret that as:

```html
{{-- Snippet 8: resources/views/livewire/hello-world/markdown-editor.blade.php --}}
<div>
    <div>
        <textarea wire:model="text" id="my-text-area"></textarea>
    </div>
    <div>
        {{ $text }}
    </div>

    <script>
        document.addEventListener('livewire:load', function () {
            const easyMDE = new EasyMDE({
                element: document.getElementById('my-text-area'),
                forceSync: true
            });
        });
    </script>
</div>
```

It might seem stupid to interpret the documentation so literally, but it does work for the rest of the documentation which very well written. Now the editor and the text are not in sync anymore, even though the component is in sync with the textarea.

![68747470733a2f2f666f72756d2e6c61726176656c2d6c697665776972652e636f6d2f75706c6f6164732f64656661756c742f6f726967696e616c2f32582f632f636636613235313765353839393033393432616131613666396231373436646263386335373632372e706e67](https://user-images.githubusercontent.com/5369819/97091371-e6f4f380-1611-11eb-93b4-9be7dc6cd9fe.png)

Fair enough, the documentation tells us to directly [access the component instance](https://laravel-livewire.com/docs/2.x/inline-scripts). So we'd have:

```html
{{-- Snippet 9: resources/views/livewire/hello-world/markdown-editor.blade.php --}}
<div>
    <div>
        <textarea wire:model="text" id="my-text-area"></textarea>
    </div>
    <div>
        {{ $text }}
    </div>

    <script>
        document.addEventListener('livewire:load', function () {
            const easyMDE = new EasyMDE({
                element: document.getElementById('my-text-area'),
                forceSync: true
            });

            easyMDE.codemirror.on("change", function(){
                @this.text = easyMDE.value();
            });
        });
    </script>
</div>
```

Now the markdown editor becomes a simple textarea whenever I type something:

![](https://forum.laravel-livewire.com/uploads/default/original/2X/8/81fa54cca6180723ea3816c7ee63e6f67a0503f0.png)
![68747470733a2f2f666f72756d2e6c61726176656c2d6c697665776972652e636f6d2f75706c6f6164732f64656661756c742f6f726967696e616c2f32582f382f383166613534636361363138303732336561333831366337656536336536663637613035303366302e706e67](https://user-images.githubusercontent.com/5369819/97091374-eb211100-1611-11eb-9c2f-b31d8a87e0ed.png)

## Ignoring DOM-changes

That's as far as the documentation for inline javascript goes. It's a shame this section is not as complete as the AlpineJS section, for instance. And that's when the workarounds begin. Somewhere else, the documentation tells us we can ignore these updates with `wire:ignore` so that doesn't happen. So we have:

```html
{{-- Snippet 10: resources/views/livewire/hello-world/markdown-editor.blade.php --}}
<div>
    <div wire:ignore>
        <textarea wire:model="text" id="my-text-area"></textarea>
    </div>
    <div>
        {{ $text }}
    </div>

    <script>
        document.addEventListener('livewire:load', function () {
            const easyMDE = new EasyMDE({
                element: document.getElementById('my-text-area'),
                forceSync: true
            });

            easyMDE.codemirror.on("change", function(){
                @this.text = easyMDE.value();
            });
        });
    </script>
</div>
```

And the parent component works now. However, this is a documentation problem for newbies like me, and all libraries should support newbies to some extent or there will never be a community. The documentation on [Inline Scripts](https://laravel-livewire.com/docs/2.x/inline-scripts) says nothing about `wire:ignore`. 

The problem here is that `wire:ignore` is mentioned in the [AlpineJS Integration section](https://laravel-livewire.com/docs/2.x/alpine-js) of the documentation. And only by the end (line 216 of 333). Even worse in my case, this is one of the few sections I didn't read at first. It might seem stupid but the rationale was that I don't work with AlpineJS so there's no reason to worry about details in the last sections on how to integrate livewire with AlpineJS.

As a newbie, it wasn't clear to me that understanding AlpineJS was that important to Livewire, no matter whether I work with AlpineJS. I don't know if that's intentional. It seemed unreasonable to have 367 lines of documentation for integration with AlpineJS, which is a javascript library, and 44 lines of documentation for "integration" with javascript itself. In fact, the AlpineJS Integration section includes a complete example that resembles pretty much what I'm trying to achieve here for Inline Javascript.

## Nested components

In any case, with the `wire:ignore` issue out of the way, we can consider the nested components, which don't work yet:

![68747470733a2f2f666f72756d2e6c61726176656c2d6c697665776972652e636f6d2f75706c6f6164732f64656661756c742f6f726967696e616c2f32582f312f313338626162396434363030656432363633613466343566616664616635393339386162323937342e706e67](https://user-images.githubusercontent.com/5369819/97091383-f7a56980-1611-11eb-8799-870e89636cfe.png)

At this point, the documentation didn't help me much anymore and it's when I started looking on github issues and the [livewire forum](https://forum.laravel-livewire.com). It's also when I started trigerring all kinds of js events trying to make this work.

Once we include the nested components, the results are just arbitrary. First of all, we have the same id for the editors and that's not livewire's fault, so we need an extra field to fix that:

```html
{{-- Snippet 11: resources/views/helloworld.blade.php --}}
{{-- ... --}}
    <livewire:hello-world.markdown-editor key="parent_editor" element-id="parent_markdown_editor"/>
{{-- ... --}}
```

```html
{{-- Snippet 12: resources/views/livewire/hello-world/nested-markdown-editor.blade.php --}}
<div>
    <div>
        <livewire:hello-world.markdown-editor key="nested_editor_1" element-id="nested_markdown_editor_1"/>
    </div>
    <div wire:init="loadEditor">
        @if ($readyToLoad)
            <livewire:hello-world.markdown-editor key="nested_editor_2" element-id="nested_markdown_editor_2"/>
        @endif
    </div>
</div>
```

```php
<?php
// Snippet 13: app/Http/Livewire/HelloWorld/MarkdownEditor.php

namespace App\Http\Livewire\HelloWorld;

use Livewire\Component;

class MarkdownEditor extends Component
{

    public $elementId;
    public $text = "# Header\nInitial Markdown";

    public function mount($elementId) {
        $this->elementId = $elementId;
    }

    public function render()
    {
        return view('livewire.hello-world.markdown-editor');
    }
}
```

```php
{{-- Snippet 14: resources/views/livewire/hello-world/markdown-editor.blade.php --}}
<div>
    <div wire:ignore>
        <textarea wire:model="text" id="{{ $elementId }}"></textarea>
    </div>
    <div>
        {{ $text }}
    </div>

    <script>
        document.addEventListener('livewire:load', function () {
            const easyMDE = new EasyMDE({
                element: document.getElementById('{{ $elementId }}'),
                forceSync: true
            });

            easyMDE.codemirror.on("change", function(){
                @this.text = easyMDE.value();
            });
        });
    </script>
</div>
```

Now we expect things to work, but the nested component that is rendered later does not work:

![68747470733a2f2f666f72756d2e6c61726176656c2d6c697665776972652e636f6d2f75706c6f6164732f64656661756c742f6f726967696e616c2f32582f382f383134363861396133306465326230376331343366373536383930386133636564633435633564342e706e67](https://user-images.githubusercontent.com/5369819/97091386-fffda480-1611-11eb-8c28-574173cc5bc3.png)

## Nested components rendered later

So the nested component does not work anymore. And here it's when things get ugly. I was sort of disappointed because the documentation had been so great up to this section.

If the nested component will be rendered later, the `addEventListener` is not useful because `livewire:load` will not be fired again. Maybe there's another event we should listen in this case, but `livewire:load` is the only event described in the [Inline Scripts Section](https://laravel-livewire.com/docs/2.x/inline-scripts) and it's not described at all in the [Reference Section](https://laravel-livewire.com/docs/2.x/reference). So I assume it's an ad hoc event just for that. I'm not sure about that. 

What we can do is run the script directly when the component is rendered later. If you understand the internals of the library, there's probably a much more elegant way of doing it (help needed here) but we can do:

```html
{{-- Snippet 15: resources/views/livewire/hello-world/markdown-editor.blade.php --}}
<div>
    <div wire:ignore>
        <textarea wire:model="text" id="{{ $elementId }}"></textarea>
    </div>
    <div>
        {{ $text }}
    </div>

    <script>
        @if (!$renderedLater)
        document.addEventListener('livewire:load', function () {
            @endif
            const easyMDE = new EasyMDE({
                element: document.getElementById('{{ $elementId }}'),
                forceSync: true
            });

            easyMDE.codemirror.on("change", function () {
            @this.text = easyMDE.value();
            });
            @if (!$renderedLater)
        });
        @endif
    </script>
</div>
```

```php
<?php
// Snippet 16: app/Http/Livewire/HelloWorld/MarkdownEditor.php

namespace App\Http\Livewire\HelloWorld;

use Livewire\Component;

class MarkdownEditor extends Component
{

    public $elementId;
    public $text = "# Header\nInitial Markdown";
    public $renderedLater;

    public function mount($elementId, $renderedLater = false) {
        $this->elementId = $elementId;
        $this->renderedLater = $renderedLater;
    }

    public function render()
    {
        return view('livewire.hello-world.markdown-editor');
    }
}
```

```html
{{-- Snippet 17: resources/views/livewire/hello-world/nested-markdown-editor.blade.php --}}
<div>
    <div>
        <livewire:hello-world.markdown-editor key="nested_editor_1" element-id="nested_markdown_editor_1"/>
    </div>
    <div wire:init="loadEditor">
        @if ($readyToLoad)
            <livewire:hello-world.markdown-editor key="nested_editor_2" element-id="nested_markdown_editor_2" rendered-later/>
        @endif
    </div>
</div>
```

And now it all works :partying_face: : 

![68747470733a2f2f666f72756d2e6c61726176656c2d6c697665776972652e636f6d2f75706c6f6164732f64656661756c742f6f726967696e616c2f32582f362f363062336435343431643161333038633366633938376161643061373464303932636264323864312e706e67](https://user-images.githubusercontent.com/5369819/97091388-0855df80-1612-11eb-894c-87cc3701986c.png)

## Livewire loading again

Although this simple example works, this solution is still not robust enough. For instance, there are cases when livewire:load is triggered many times and this can cause problems. Consider an application where this happens. I include

```html
<livewire:hello-world.markdown-editor key="parent_editor" element-id="editor"/>
```

And I get:

![68747470733a2f2f666f72756d2e6c61726176656c2d6c697665776972652e636f6d2f75706c6f6164732f64656661756c742f6f726967696e616c2f32582f652f656339366566303261373539393934663163643230663865323966663833373938363538633066642e706e67](https://user-images.githubusercontent.com/5369819/97091392-0ee45700-1612-11eb-9bdb-136c0cfdc09f.png)

And a few more. Even worse, none these editors update their `$text` variable for some reason. I don't even know why `livewire:load` is being triggered so many times, if I should be listening to another event, or even if there's no solution in this direction. `livewire:load` is not in the [Reference section](https://laravel-livewire.com/docs/2.x/reference) and I'm figuring all this out by myself.

If we were to come up with an ugly workaround, we could prohibit the initialization function to run more than once:

```html
<script>
        function markdownInit{{ $elementId }}() {
            if ( typeof markdownInit{{ $elementId }}.initialized == 'undefined' ) {
                markdownInit{{ $elementId }}.initialized = false;
            }

            if (!markdownInit{{ $elementId }}.initialized) {
                console.log('Markdown init called');
                const easyMDE = new EasyMDE({
                    element: document.getElementById('{{ $elementId }}'),
                    forceSync: true
                });

                easyMDE.codemirror.on("change", function () {
                    @this.text = easyMDE.value();
                });
            }

            markdownInit{{ $elementId }}.initialized = true;
        }

        @if (!$renderedLater)
        document.addEventListener('livewire:load', function () {
            @endif
                console.log('Livewire:load called');
                markdownInit{{ $elementId }}();
            @if (!$renderedLater)
        });
        @endif
    </script>
```

And now we get

![68747470733a2f2f666f72756d2e6c61726176656c2d6c697665776972652e636f6d2f75706c6f6164732f64656661756c742f6f726967696e616c2f32582f352f353163336637343963343964363565643533643461306236326335633164653563643838663335332e706e67](https://user-images.githubusercontent.com/5369819/97091398-1441a180-1612-11eb-9488-0693eaa5f894.png)

but, as you can see, the `$text` also doesn't get updated (help needed here).

## Overall

Well, I hope this helps someone. It's very limited but it's more information about about inline javascript than I could find anywhere else. I'm not very experienced with livewire or javascript, but it would be good if we can at least make sure some other newbies like me don't have to go through all that by themselves. 

If you are more experienced, maybe you can point out more elegant ways to identify elements that are loaded later, might know to which other events we should be listening, or how to fix the limitation I described in the last section. I'll update this post as needed.

Best,

**Comment by calebporzio** [comment]
Wow, great writeup! Thanks for taking the time!

**Reply by alandefreitas** [reply]
Thanks!

**Comment by mokhosh** [comment]
I guess instead of that horrible Blade JS mashup, you can put this in your template:
```js
window.mdeInit = (id) => {
  if (window.mde && window.mde[id]) return
  
  const easyMDE = new EasyMDE({
    element: document.getElementById(id),
    forceSync: true
  }) 
  easyMDE.codemirror.on('change', () => @this.text = easyMDE.value())
  
  window.mde || window.mde = []
  window.mde[id] = true
}
```
and put this in your component view:
```js
if (window.livewire) mdeInit('{{ $elementId }}')
else document.addEventListener('livewire:load', () => mdeInit('{{ $elementId }}'))
```


And get rid of all of these:
```php
public $renderedLater;

public function mount($renderedLater = false) {
    $this->renderedLater = $renderedLater;
}
```
```html
<script>
  function markdownInit{{ $elementId }}() {
      if ( typeof markdownInit{{ $elementId }}.initialized == 'undefined' ) {
          markdownInit{{ $elementId }}.initialized = false;
      }

      if (!markdownInit{{ $elementId }}.initialized) {
          console.log('Markdown init called');
          const easyMDE = new EasyMDE({
              element: document.getElementById('{{ $elementId }}'),
              forceSync: true
          });

          easyMDE.codemirror.on("change", function () {
              @this.text = easyMDE.value();
          });
      }

      markdownInit{{ $elementId }}.initialized = true;
  }

  @if (!$renderedLater)
  document.addEventListener('livewire:load', function () {
      @endif
          console.log('Livewire:load called');
          markdownInit{{ $elementId }}();
      @if (!$renderedLater)
  });
  @endif
</script>
```
Right?
